File: modules/tests/src/commonTest/kotlin/primitives/CallbackArgPrimitive.kt
package primitives

import natives.test.MyDictionary
import natives.test.MyEnum
import natives.test.VoidCallback
import withLib
import kotlin.test.Test
import kotlin.test.assertTrue

class CallbackArgPrimitive {

    @Test
    fun callbackVoid() = withLib {
        var pass = false
        natives.test.callbackVoid { pass = true }
        assertTrue(pass)
    }

    @Test
    fun callbackVoidN() = withLib {
        assertTrue(natives.test.callbackVoidN(null))
    }

    @Test
    fun callbackArgChar() = withLib {
        assertTrue(natives.test.callbackArgChar { it == 'a' })
    }

    @Test
    fun callbackArgBoolean() = withLib {
        assertTrue(natives.test.callbackArgBoolean { it })
    }

    @Test
    fun callbackArgByte() = withLib {
        assertTrue(natives.test.callbackArgByte { it == 1.toByte() })
    }

    @Test
    fun callbackArgUByte() = withLib {
        assertTrue(natives.test.callbackArgUByte { it == UByte.MAX_VALUE })
    }

    @Test
    fun callbackArgShort() = withLib {
        assertTrue(natives.test.callbackArgShort { it == 1.toShort() })
    }

    @Test
    fun callbackArgUShort() = withLib {
        assertTrue(natives.test.callbackArgUShort { it == UShort.MAX_VALUE })
    }

    @Test
    fun callbackArgInt() = withLib {
        assertTrue(natives.test.callbackArgInt { it == 1 })
    }

    @Test
    fun callbackArgUInt() = withLib {
        assertTrue(natives.test.callbackArgUInt { it == UInt.MAX_VALUE })
    }

    @Test
    fun callbackArgLong() = withLib {
        assertTrue(natives.test.callbackArgLong { it == 1.toLong() })
    }

    @Test
    fun callbackArgULong() = withLib {
        assertTrue(natives.test.callbackArgULong { it == ULong.MAX_VALUE })
    }

    @Test
    fun callbackArgFloat() = withLib {
        assertTrue(natives.test.callbackArgFloat { it == 1.1f })
    }

    @Test
    fun callbackArgDouble() = withLib {
        assertTrue(natives.test.callbackArgDouble { it == 1.1 })
    }

    @Test
    fun callbackArgString() = withLib {
        assertTrue(natives.test.callbackArgString { it == "test string" })
    }

    @Test
    fun callbackArgStringN() = withLib {
        assertTrue(natives.test.callbackArgStringN { it == null })
    }

    @Test
    fun callbackArgCallback() = withLib {
        val toPass = VoidCallback {}
        assertTrue(natives.test.callbackArgCallback(toPass) { it == toPass })
    }

    @Test
    fun callbackArgCallbackN() = withLib {
        assertTrue(natives.test.callbackArgCallbackN { it == null })
    }

    @Test
    fun callbackArgEnum() = withLib {
        assertTrue(natives.test.callbackArgEnum { it == MyEnum.CASE2 })
    }

    @Test
    fun callbackArgDictionary() = withLib {
        assertTrue(natives.test.callbackArgDictionary { it == MyDictionary(1, 2, 3, 4) })
    }

    @Test
    fun callbackArgDictionaryN() = withLib {
        assertTrue(natives.test.callbackArgDictionaryN { it == null })
    }
}

File: modules/tests/natives/testcpp/src/primitives/callback_arg_primitive.cpp
#include <api.hpp>

// Consume

void callback_void(std::shared_ptr<VoidCallback> arg) {
    arg->invoke();
}

bool callback_void_n(KOptional<std::shared_ptr<VoidCallback>> arg) {
    return arg.is_none();
}

bool callback_arg_char(std::shared_ptr<CallbackPassChar> arg) {
    return arg->invoke('a');
}

bool callback_arg_boolean(std::shared_ptr<CallbackPassBoolean> arg) {
    return arg->invoke(true);
}

bool callback_arg_byte(std::shared_ptr<CallbackPassByte> arg) {
    return arg->invoke(1);
}

bool callback_arg_ubyte(std::shared_ptr<CallbackPassUByte> arg) {
    return arg->invoke(255u);
}

bool callback_arg_short(std::shared_ptr<CallbackPassShort> arg) {
    return arg->invoke(1);
}

bool callback_arg_ushort(std::shared_ptr<CallbackPassUShort> arg) {
    return arg->invoke(65535u);
}

bool callback_arg_int(std::shared_ptr<CallbackPassInt> arg) {
    return arg->invoke(1);
}

bool callback_arg_uint(std::shared_ptr<CallbackPassUInt> arg) {
    return arg->invoke(4294967295u);
}

bool callback_arg_long(std::shared_ptr<CallbackPassLong> arg) {
    return arg->invoke(1);
}

bool callback_arg_ulong(std::shared_ptr<CallbackPassULong> arg) {
    return arg->invoke(18446744073709551615u);
}

bool callback_arg_float(std::shared_ptr<CallbackPassFloat> arg) {
    return arg->invoke(1.1f);
}

bool callback_arg_double(std::shared_ptr<CallbackPassDouble> arg) {
    return arg->invoke(1.1);
}

bool callback_arg_string(std::shared_ptr<CallbackPassString> arg) {
    return arg->invoke(KString("test string"));
}

bool callback_arg_string_n(std::shared_ptr<CallbackPassStringN> arg) {
    return arg->invoke(KOptional<KString>());
}

bool callback_arg_callback(std::shared_ptr<VoidCallback> pass, std::shared_ptr<CallbackPassCallback> arg) {
    return arg->invoke(pass);
}

bool callback_arg_callback_n(std::shared_ptr<CallbackPassCallbackN> arg) {
    return arg->invoke(KOptional<std::shared_ptr<VoidCallback>>());
}

bool callback_arg_enum(std::shared_ptr<CallbackPassEnum> arg) {
    return arg->invoke(CASE2);
}

bool callback_arg_dictionary(std::shared_ptr<CallbackPassDictionary> arg) {
    return arg->invoke(MyDictionary(1, 2, 3, 4));
}

bool callback_arg_dictionary_n(std::shared_ptr<CallbackPassDictionaryN> arg) {
    return arg->invoke(KOptional<MyDictionary>());
}

bool callback_arg_interface(std::shared_ptr<IMyInterface> pass, std::shared_ptr<CallbackPassInterface> arg) {
    return arg->invoke(std::shared_ptr(pass));
}

bool callback_arg_interface_n(std::shared_ptr<CallbackPassInterfaceN> arg) {
    return arg->invoke(KOptional<std::shared_ptr<IMyInterface>>());
}
